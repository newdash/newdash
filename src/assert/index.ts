import isAsyncFunction from "../isAsyncFunction";
import isDate from "../isDate";
import type { AdvancedType, JSType } from "../types";

/**
 * must provide value with type
 *
 * @param value
 * @param fieldName
 * @param type
 */
export function mustProvide(value: any, fieldName: string, type: JSType) {

  if (value === undefined || value === null || typeof value !== type) {
    throw new TypeError(`must provide '${fieldName}' value with type '${type}', given: '${value}'/'${type}'`);
  }

}


/**
 * must provide some value
 *
 * @param value
 * @param fieldName
 * @param type
 * @returns
 */
export function mustProvideN(value: any, fieldName: string, type: AdvancedType) {
  if (
    value === undefined ||
    value === null ||
    (type === "async_function" && !isAsyncFunction(value)) ||
    (type === "date" && !isDate(value))
  ) {
    throw new TypeError(`must provide '${fieldName}' value with type '${type}', given: '${value}'/'${typeof value}'`);
  }

}
