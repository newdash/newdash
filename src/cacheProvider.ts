import { LRUMap } from "./functional/LRUMap";
import { TTLMap } from "./functional/TTLMap";
import { GeneralFunction } from "./types";

export interface CachePolicy {
  cacheUndefined?: boolean;
  cacheNull?: boolean;
  cacheThrow?: boolean;
}

export interface CacheConfig<T> {
  policy?: CachePolicy;
  params?: T;
}

export type LRUCacheConfig = CacheConfig<LRUCacheProviderParam>;

export type TTLCacheConfig = CacheConfig<TTLCacheProviderParam>;

const DEFAULT_CACHE_POLICY: CachePolicy = {
  cacheUndefined: false,
  cacheNull: false,
  cacheThrow: false
};

class CachedThrowError {
  private readonly _error: Error;
  constructor(error: Error) { this._error = error; }
  getError() { return this._error; }
}

export interface CacheProvider<K, V> extends Map<K, V> {

  /**
   * get cache or create it on necessary
   *
   * @param key
   * @param producer
   */
  getOrCreate<R>(key: K, producer: GeneralFunction<any[], R>): R

}

/**
 * async cache provider
 */
export interface AsyncCacheProvider<K, V> extends CacheProvider<K, Promise<V>> { }

export interface LRUCacheProviderParam {
  maxEntry: number;
}

function defaultGetOrCreate(cache: Map<any, any>, policy: CachePolicy, key: any, producer: any) {
  if (!cache.has(key)) {
    try {
      const value = producer();
      // work with async function
      if (value instanceof Promise) {
        // @ts-ignore
        return value
          .then((result) => {
            if (
              result === null && !Boolean(policy.cacheNull) ||
              result === undefined && !Boolean(policy.cacheUndefined)
            ) {
              // do nothing
            } else {
              cache.set(key, result);
            }
            return result;
          })
          .catch((error) => {
            if (Boolean(policy.cacheThrow)) {
              cache.set(key, new CachedThrowError(error));
            }
            throw error;
          });
      }
      if (
        value === null && !Boolean(policy.cacheNull) ||
        value === undefined && !Boolean(policy.cacheUndefined)
      ) {
        // do nothing
      } else {
        cache.set(key, value);
      }
      return value;
    } catch (error) {
      if (Boolean(policy.cacheThrow)) {
        cache.set(key, new CachedThrowError(error));
      }
      throw error;
    }

  }
  const cachedValue = cache.get(key);
  if (cachedValue instanceof CachedThrowError) {
    throw cachedValue.getError();
  }
  return cachedValue;
}

/**
 * LRU Cache Provider
 *
 * @category Cache
 * @since 5.16.0
 */
export class LRUCacheProvider<K = any, V = any> extends LRUMap implements CacheProvider<K, V> {

  protected readonly _cachePolicy: CachePolicy = DEFAULT_CACHE_POLICY;

  constructor(config: CacheConfig<LRUCacheProviderParam>)
  constructor(maxEntry?: number)
  constructor(param0: any) {
    super(typeof param0 === "number" ? param0 : (param0?.params?.maxEntry ?? 10240));
    if (typeof param0 === "object") {
      this._cachePolicy = Object.assign(DEFAULT_CACHE_POLICY, param0.policy ?? {});
    }
  }

  public getOrCreate<R>(key: K, producer: GeneralFunction<any[], R>): R {
    return defaultGetOrCreate(this, this._cachePolicy, key, producer);
  }

}

export interface TTLCacheProviderParam {
  ttl?: number;
  checkInterval?: number;
  maxEntry?: number;
}

const DEFAULT_CACHE_PROVIDER_PARAM: TTLCacheProviderParam = {
  ttl: 30 * 1000,
  checkInterval: 60 * 1000,
  maxEntry: 10240
};

/**
 * TTL Cache Provider
 *
 * @since 5.16.0
 * @category Cache
 *
 */
export class TTLCacheProvider<K = any, V = any> extends TTLMap<K, V> implements CacheProvider<K, V> {

  protected readonly _cachePolicy: CachePolicy;

  constructor(config: CacheConfig<TTLCacheProviderParam>)
  constructor(ttl?: number, checkInterval?: number, maxEntry?: number)
  constructor(...params: any[]) {
    const config = {
      policy: DEFAULT_CACHE_POLICY, params: DEFAULT_CACHE_PROVIDER_PARAM
    };
    if (typeof params[0] === "number") {
      config.params.ttl = params[0] || config.params.ttl;
      config.params.checkInterval = params[1] || config.params.checkInterval;
      config.params.maxEntry = params[2] || config.params.maxEntry;
    } else {
      config.policy = Object.assign(config.policy || {}, params[0].policy || {});
      config.params = Object.assign(config.params || {}, params[0].params || {});
    }
    super(config.params.ttl);
    this._cachePolicy = config.policy;
  }

  public getOrCreate<R>(key: K, producer: GeneralFunction<any[], R>): R {
    return defaultGetOrCreate(this, this._cachePolicy, key, producer);
  }

}

export const cacheProvider = {
  LRUCacheProvider,
  TTLCacheProvider
};
