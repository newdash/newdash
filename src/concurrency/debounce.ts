import { mustProvide } from "../assert";
import { debounce as nativeDebounce } from "../debounce";
import { defineFunctionName } from "../functional/defineFunctionName";
import { AsyncFunction } from "../types";


/**
 * concurrency simply debounce function
 *
 * it will create a debounced function that delays invoking func until after wait milliseconds have elapsed since the last time the debounced function was invoked
 *
 * @since 5.18.0
 * @category Async
 *
 * @param runner
 * @param wait wait milliseconds before last time invocation
 */
export function debounce<T extends AsyncFunction>(runner: T, wait: number): T {
  mustProvide(runner, "runner", "function");
  mustProvide(wait, "wait", "number");
  return defineFunctionName(nativeDebounce(runner, wait), runner?.name) as any as T;
}


export default debounce;
