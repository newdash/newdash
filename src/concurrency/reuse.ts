import { mustProvide, mustProvideN } from "../assert";
import { createFunctionWrapper } from "../functional/functionWrapper";
import LRUMap from "../functional/LRUMap";
import toHashCode from "../functional/toHashCode";
import { AsyncFunction } from "../types";


/**
 * reuse values in specific duration for async functions
 *
 * @category Async
 * @since 5.22.0
 * @param runner must be an async function
 * @param duration default 1000 milliseconds
 * @param maxHandles different cache values for parameters
 * @returns
 */
export function reuse<T extends AsyncFunction>(runner: T, duration: number = 1000, maxHandles = 1000) {

  mustProvideN(runner, "runner", "async_function");
  mustProvide(duration, "duration", "number");
  mustProvide(maxHandles, "maxHandles", "number");

  return createFunctionWrapper(runner, {
    global: {
      duration,
      handles: new LRUMap<string, { value: any, timeout: number }>(maxHandles),
    },
    execute(ctx) {
      const { args, global: { handles } } = ctx;
      const key = toHashCode(args);
      if (handles.has(key)) {
        const { value, timeout } = ctx.global.handles.get(key);
        if (timeout >= Date.now()) { return value; }
      }
      const v = ctx.runner(...ctx.args);
      handles.set(key, { value: v, timeout: Date.now() + duration });
      return v;
    },
  });

}

export default reuse;
