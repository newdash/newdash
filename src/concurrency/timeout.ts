import { mustProvide } from "../assert";
import defineFunctionName from "../functional/defineFunctionName";
import { createTimeoutPromise } from "../timeout";
import { AsyncFunction } from "../types";

/**
 * wrap an async function with timeout
 *
 * @since 5.15.0
 * @category Async
 *
 * @param runner async runner please, otherwise the `timeout` is not meaningful
 * @param timeout timeout threshold in milliseconds, default value is 60 seconds
 *
 * @throws {TimeoutError}
 *
 * @example
 *
 * ```ts
 * const f = timeout(async() => {}, 1000)
 * // f() will throw error if the result is not resolved in one second
 * ```
 *
 */
export function timeout<T extends AsyncFunction>(runner: T, timeout: number = 60 * 1000): T {

  mustProvide(runner, "runner", "function");
  mustProvide(timeout, "timeout", "number");

  const func = (...args: any[]) => createTimeoutPromise(async (resolve, reject) => {
    try {
      resolve(await runner(...args));
    } catch (error) {
      reject(error);
    }
  }, timeout);

  return defineFunctionName(func, runner.name) as any;
}

export default timeout;
