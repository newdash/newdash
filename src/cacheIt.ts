/* eslint-disable max-len */
import { CacheConfig, CacheProvider, LRUCacheProvider } from "./cacheProvider";
import { toHashCode } from "./functional/toHashCode";
import { isClass } from "./isClass";
import { Class, GeneralFunction } from "./types";

/**
 * @private
 * @ignore
 * @internal
 */
const KEY_CACHE_PROPERTY = "__cache_storage";
/**
 * @private
 * @ignore
 * @internal
 */
const KEY_CLEAR_CACHE_FUNCTION = "__cache_clear";

export type CachedFunction<T extends GeneralFunction> = T & {
  /**
   * get the cache provider
   */
  [KEY_CACHE_PROPERTY]: CacheProvider<string, ReturnType<T>>;
  /**
   * clear cache by parameters
   */
  [KEY_CLEAR_CACHE_FUNCTION]: (...args: Parameters<T>) => void;
}

export type CachedObject<T = any> = {
  [K in keyof T]: T[K] extends (...args: any[]) => any | T[K] ? CachedFunction<T[K]> : T[K]
}

export type CachedClass<T extends new (...args: any[]) => any> = new (...args: any[]) => CachedObject<InstanceType<T>>

export interface CacheItOptions<CACHE_PARAM = any> {
  /**
   * provider type, the constructor of CacheProvider
   */
  provider?: Class<CacheProvider<any, any>>;
  /**
   * provider constructor args
   *
   * @deprecated
   */
  providerArgs?: Array<any>;
  /**
   * provider config
   */
  providerOptions?: CacheConfig<CACHE_PARAM>;
}

const defaultCacheItOptions: CacheItOptions = {
  provider: LRUCacheProvider,
  providerArgs: undefined,
  providerOptions: {},
};

/**
 * @private
 * @ignore
 * @internal
 * @param obj
 * @param options
 */
function cacheItFunction(obj: any, options?: CacheItOptions) {
  const cacheProvider = options.providerOptions !== undefined ? new options.provider(options.providerOptions) : new options.provider(...options.providerArgs);

  const clearCache = (...args: any[]) => {
    cacheProvider.delete(toHashCode(args));
  };

  return new Proxy(obj, {
    apply: (target, ctx, args) => {
      const producer = () => target.apply(ctx, args);
      const key = toHashCode(args);
      return cacheProvider.getOrCreate(key, producer);
    },
    get: (target, propertyName) => {
      if (propertyName in target) {
        return target[propertyName];
      }
      if (propertyName === KEY_CACHE_PROPERTY) {
        return cacheProvider;
      }
      if (propertyName === KEY_CLEAR_CACHE_FUNCTION) {
        return clearCache;
      }
      return undefined;
    }
  });

}

/**
 * @private
 * @ignore
 * @internal
 * @param obj
 * @param options
 */
function cacheItClass<T>(obj: Class<T>, options?: CacheItOptions) {

  return new Proxy(obj, {
    construct: (target, args) => cacheItObject(new target(...args), options)
  });

}

/**
 * @private
 * @ignore
 * @internal
 * @param obj
 * @param options
 */
function cacheItObject(obj: any, options?: CacheItOptions) {

  const methodsCacheProvider = new LRUCacheProvider(10240);

  return new Proxy(obj, {
    get: ((target, propertyName) => {
      if (propertyName in target) {
        const propertyValue = target[propertyName];
        if (typeof propertyValue === "function") {
          const producer = () => cacheItFunction(propertyValue, options);
          return methodsCacheProvider.getOrCreate(propertyName, producer);
        }
        return propertyValue;
      }
    })
  });

}

/**
 * make all methods of object are cached
 *
 * @since 5.16.0
 * @category Util
 * @param obj
 * @param options
 * @example
 *
 * ```ts
 *
    // set max cache size
    const f = cacheIt((value: number) => (++idx) + value, {
      providerOptions: {
        params: {
          max_entries: 2
        }
      }
    });
    expect(f.__cache_storage["_maxSize"]).toBe(2);

    expect(f(0)).toBe(1);
    expect(f(0)).toBe(1);
    ```
 */
export function cacheIt<T extends GeneralFunction>(obj: T, options?: CacheItOptions): CachedFunction<T>
/**
 * create a class wrapper, the instance created by that wrapper will automatically apply `cacheIt`
 * @since 5.16.0
 * @param obj
 * @param options
 *
 * @example
 *
 * ```ts
 *  class A {
      constructor(idx = 0) { this.idx = idx; }
      private idx: number
      public add(value: number) {
        return (++this.idx) + value;
      }
    }
    const CachedA = cacheIt(A);
    const a = new CachedA(1);
    expect(a.add(0)).toBe(2);
    expect(a.add(0)).toBe(2);

    expect(a.add(1)).toBe(4);
    expect(a.add(1)).toBe(4);
  ```
 */
export function cacheIt<T extends Class>(obj: T, options?: CacheItOptions): CachedClass<T>
/**
 * make function is cached, default with LRU container
 *
 * @since 5.16.0
 * @category Util
 * @param obj
 * @param options
 */
export function cacheIt<T extends object>(obj: T, options?: CacheItOptions): CachedObject<T>
export function cacheIt<T extends object>(obj: T, options?: CacheItOptions): CachedObject<T>
export function cacheIt(obj: any, options?: any): any {

  // assign default option
  options = Object.assign({}, defaultCacheItOptions, options);

  switch (typeof obj) {
    case "object":
      return cacheItObject(obj, options);
    case "function":
      if (isClass(obj)) {
        return cacheItClass(obj, options);
      }
      return cacheItFunction(obj, options);
    default:
      return obj;
  }

}


export default cacheIt;
