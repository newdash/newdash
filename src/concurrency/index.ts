import { any } from "../any";
import { series } from "../series";
import { debounce } from "./debounce";
import { limit, synchronized } from "./limit";
import { reuse } from "./reuse";
import { timeout } from "./timeout";


/**
 * concurrency functions
 */
export const concurrency = {
  limit,
  timeout,
  any,
  series,
  debounce,
  synchronized,
  reuse,
};
