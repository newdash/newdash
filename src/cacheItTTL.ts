/* eslint-disable max-len */
import cacheIt, { CachedClass, CachedFunction, CachedObject, CacheItOptions } from "./cacheIt";
import { TTLCacheProvider, TTLCacheProviderParam } from "./cacheProvider";
import { Class, GeneralFunction } from "./types";

export function cacheItTTL<T extends GeneralFunction>(obj: T, options?: CacheItOptions<TTLCacheProviderParam>): CachedFunction<T>
export function cacheItTTL<T extends Class>(obj: T, options?: CacheItOptions<TTLCacheProviderParam>): CachedClass<T>
export function cacheItTTL<T extends object>(obj: T, options?: CacheItOptions<TTLCacheProviderParam>): CachedObject<T>
export function cacheItTTL(obj: any, options?: CacheItOptions<TTLCacheProviderParam>): any
export function cacheItTTL(obj: any, options: any = {}) {
  const opt: CacheItOptions = Object.assign({}, { provider: TTLCacheProvider }, options);
  return cacheIt(obj, opt);
}

