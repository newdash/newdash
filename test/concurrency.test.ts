// @ts-nocheck
import { any, sleep } from "../src";
import { concurrency } from "../src/concurrency";
import { assertShouldThrowError } from "./helpers";
describe("concurrency", () => {

  it("should support concurrency.limit function", async () => {

    const currLimit = 5;
    let ctx = 0;

    const runner = concurrency.limit(async () => ctx++, currLimit);

    const allRunner = [];

    for (let idx = 0; idx < 15; idx++) {
      allRunner.push(runner().then((v) => {
        expect(ctx).toBeLessThanOrEqual(currLimit);
        ctx--;
        return v;
      }));
    }

    const values = await Promise.all(allRunner);
    expect(values).toHaveLength(15);
    expect(values).not.toBeUndefined();
  });

  it("should support concurrency.limit function on error", async () => {

    class E1 extends Error { }

    const currLimit = 5;
    let ctx = 0;
    let total = 0;
    const max = 15;

    const runner = concurrency.limit(async () => {
      ctx++;
      total++;
      if (total > (max - 2)) {
        throw new E1();
      }
    }, currLimit);

    const allRunner = [];

    for (let idx = 0; idx < max; idx++) {
      allRunner.push(runner().then(() => {
        expect(ctx).toBeLessThanOrEqual(currLimit);
        ctx--;
      }));
    }
    await assertShouldThrowError(() => Promise.all(allRunner), E1);


  });

  it('should support alias of function "any"', () => {
    expect(concurrency.any).toBe(any);
  });

  it('should validate reuse values parameter', async () => {
    expect(() => concurrency.reuse(() => { })).toThrowError()
  });

  it('should support reuse values for async function', async () => {
    const fn = async function () {
      return fn.context_value = (fn.context_value ?? 0) + 1
    }
    const reusedFn = concurrency.reuse(fn, 300)
    await expect(reusedFn()).resolves.toBe(1)
    await expect(reusedFn()).resolves.toBe(1)
    await expect(reusedFn()).resolves.toBe(1)

    await sleep(300)

    await expect(reusedFn()).resolves.toBe(2)
    await expect(reusedFn()).resolves.toBe(2)
    await expect(reusedFn()).resolves.toBe(2)
    await expect(reusedFn(1)).resolves.toBe(3)
    await expect(reusedFn('43')).resolves.toBe(4)
    await expect(reusedFn()).resolves.toBe(2)
  });

});
