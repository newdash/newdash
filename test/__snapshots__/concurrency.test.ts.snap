// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`concurrency should validate reuse values parameter 1`] = `"must provide 'runner' value with type 'async_function', given: '() => { }'/'function'"`;
